src/Main.hs

```diff
@@ -14,15 +14,143 @@
 module Main where
 
 import Vector
-
-main :: IO()
-main = print $ Circle (normalize $ Vector 1 2 3 `cross` Vector 4 5 6) 3
+import Codec.Picture
+import Codec.Picture.Png
+import Data.Maybe
+import Data.List
+import Debug.Trace
+import Data.Function
 
 -- Basic Data Types
-data Shape = Circle Vector Double
-           | Plane Vector Vector
-           | Triangle Vector Vector Vector
+data Object = Object Shape Material
+data Material = Material PixelRGB8
+data Shape = Sphere Vector Double
+--           | Plane Vector Vector 
+--           | Triangle Vector Vector Vector
             deriving (Show, Eq)
+data Light = PointLight Vector
+data Camera = Camera Double Vector Vector
+data Scene = Scene [Shape] [Light] Camera
+data Ray = Ray {origin :: Vector, direction :: Vector}
+
+sceneWidth :: Int
+sceneWidth = 500
+
+sceneHeight :: Int
+sceneHeight = 500
+
+red :: PixelRGB8
+red = PixelRGB8 255 0 0
+
+green :: PixelRGB8
+green = PixelRGB8 0 255 0
+
+blue :: PixelRGB8
+blue = PixelRGB8 0 0 255
+
+white :: PixelRGB8
+white = PixelRGB8 255 255 255
+
+black :: PixelRGB8
+black = PixelRGB8 0 0 0
+
+redMaterial :: Material
+redMaterial = Material red
+
+spheres :: [Shape]
+spheres = [Sphere (Vector (-3) 3.5 (-8)) 3, Sphere (Vector 3 3.5 (-8)) 1]
+
+objects :: [Object]
+objects = [Object 
+                (Sphere (Vector (-3) 3.5 (-8)) 3) 
+                (Material red),
+           Object
+                (Sphere (Vector 3 3.5 (-8)) 1)
+                (Material green)]
+
+lights :: [Light]
+lights = [PointLight (Vector (-30) (-10) 20) ]
+
+camera :: Camera
+camera = Camera 45 (Vector 0 1.8 10) (Vector 0 3 0)
+
+scene :: Scene
+scene = Scene spheres lights camera
+
+main :: IO()
+main = writePng "output.png" $ generateImage Main.trace sceneWidth sceneHeight
+
+minWithEmpty :: [Double] -> Double
+minWithEmpty [] = -1
+minWithEmpty list = minimum list
+
+trace x y 
```

Comment:
this should definitely have a type signature. more importantly, you can simplify it by doing the unpacking in the pattern matching. I'll chuck up a PR with an example